### PR TITLE
refactor!: Remove deprecated `Failure` crate with thiserror/anyhow/di…

### LINKED ISSUE #177: Remove deprecated `Failure` crate
We use the `Failure` crate throughout the Mun codebase but it has been deprecated. It would be a good idea to replace it.

Suggested alternatives to use in places where we do use `Failure` (which is not really a lot) are:

* [anyhow](https://github.com/dtolnay/anyhow) as a replacement for `failure::Error`
* ~~[displaydoc](https://github.com/yaahc/displaydoc) seems like a good replacement for `#[derive(Fail)]`~~
* [thiserror](https://github.com/dtolnay/thiserror) as a replacement for `#[derive(Fail)]`

**Good first issue process:**

- [ ] [read our contributing guidelines](https://github.com/mun-lang/mun/blob/master/CONTRIBUTING.md)
- [ ] claim this issue (assign yourself or comment below)
- [ ] implement the feature and start a Pull Request with in the description: closes #177. 

If this is your first PR, welcome 🎉 😄

### PR DESCRIPTION
**Pull request by attliaLin**
Close #177

### CODE REVIEW

File: crates/mun/Cargo.toml [hunk 1]
```diff
@@ -10,14 +10,16 @@ license = "MIT OR Apache-2.0"
 description = "Command-line interface for compiling, monitoring and running Mun code"
 
 [dependencies]
-failure = "0.1.7"
+anyhow = "1.0"
+thiserror = "1.0.17"
+displaydoc = "0.1"
```
Comment (Wodann) [inline crates/mun/Cargo.toml:15]: After creation of the original issue, a comment was [added](https://github.com/mun-lang/mun/issues/177#issuecomment-628609901) that proposed leaving out `displaydoc`. We agreed that that would be a better approach than using `displaydoc`. Maybe you missed that comment, but could you please remove the `displaydoc` usage?

This [article](https://nick.groenen.me/posts/rust-error-handling/) gives a really good explanation of how that can be achieved :)

File: crates/mun_codegen/src/code_gen.rs [hunk 2]
```diff
@@ -74,7 +70,7 @@ impl ObjectFile {
         })
     }
 
-    pub fn into_shared_object(self, out_dir: Option<&Path>) -> Result<PathBuf, failure::Error> {
+    pub fn into_shared_object(self, out_dir: Option<&Path>) -> Result<PathBuf> {
```
Comment (Wodann) [inline crates/mun_codegen/src/code_gen.rs:73]:
```suggestion
    pub fn into_shared_object(self, out_dir: Option<&Path>) -> Result<PathBuf, anyhow::Error> {
```

File: crates/mun_codegen/src/code_gen.rs [hunk 3]
```diff
@@ -100,7 +96,7 @@ pub struct ModuleBuilder<'a, D: IrDatabase> {
 
 impl<'a, D: IrDatabase> ModuleBuilder<'a, D> {
     /// Constructs module for the given `hir::FileId` at the specified output file location.
-    pub fn new(db: &'a D, file_id: FileId) -> Result<Self, failure::Error> {
+    pub fn new(db: &'a D, file_id: FileId) -> Result<Self> {
```
Comment (Wodann) [inline crates/mun_codegen/src/code_gen.rs:99]:
```suggestion
    pub fn new(db: &'a D, file_id: FileId) -> Result<Self, anyhow::Error> {
```

File: crates/mun_codegen/src/code_gen.rs [hunk 4]
```diff
@@ -139,7 +135,7 @@ impl<'a, D: IrDatabase> ModuleBuilder<'a, D> {
     }
 
     /// Constructs an object file.
-    pub fn build(self) -> Result<ObjectFile, failure::Error> {
+    pub fn build(self) -> Result<ObjectFile> {
```
Comment (Wodann) [inline crates/mun_codegen/src/code_gen.rs:138]:
```suggestion
    pub fn build(self) -> Result<ObjectFile, anyhow::Error> {
```

File: crates/mun_compiler/src/driver.rs [hunk 5]
```diff
@@ -50,10 +51,7 @@ impl Driver {
     }
 
     /// Constructs a driver with a configuration and a single file.
-    pub fn with_file(
-        config: Config,
-        path: PathOrInline,
-    ) -> Result<(Driver, FileId), failure::Error> {
+    pub fn with_file(config: Config, path: PathOrInline) -> Result<(Driver, FileId)> {
```
Comment (Wodann) [inline crates/mun_compiler/src/driver.rs:54]:
```suggestion
    pub fn with_file(config: Config, path: PathOrInline) -> Result<(Driver, FileId), anyhow::Error> {
```

Comment (Wodann) [reply crates/mun_compiler/src/driver.rs:54]: Beware of potential `rustfmt` errors, due to too many characters on a single line

File: crates/mun_compiler/src/driver.rs [hunk 6]
```diff
@@ -109,10 +107,7 @@ impl Driver {
 
     /// Emits all diagnostic messages currently in the database; returns true if errors were
     /// emitted.
-    pub fn emit_diagnostics(
-        &self,
-        writer: &mut dyn std::io::Write,
-    ) -> Result<bool, failure::Error> {
+    pub fn emit_diagnostics(&self, writer: &mut dyn std::io::Write) -> Result<bool> {
```
Comment (Wodann) [inline crates/mun_compiler/src/driver.rs:110]:
```suggestion
    pub fn emit_diagnostics(&self, writer: &mut dyn std::io::Write) -> Result<bool, anyhow::Error> {
```

Comment (Wodann) [reply crates/mun_compiler/src/driver.rs:110]: Beware of `rustfmt`

File: crates/mun_compiler/src/driver.rs [hunk 7]
```diff
@@ -133,7 +128,7 @@ impl Driver {
 
 impl Driver {
     /// Generate an assembly for the given file
-    pub fn write_assembly(&mut self, file_id: FileId) -> Result<PathBuf, failure::Error> {
+    pub fn write_assembly(&mut self, file_id: FileId) -> Result<PathBuf> {
```
Comment (Wodann) [inline crates/mun_compiler/src/driver.rs:131]:
```suggestion
    pub fn write_assembly(&mut self, file_id: FileId) -> Result<PathBuf, anyhow::Error> {
```

File: crates/mun_compiler/src/lib.rs [hunk 8]
```diff
@@ -57,7 +58,7 @@ impl CompilerOptions {
     }
 }
 
-pub fn main(options: CompilerOptions) -> Result<Option<PathBuf>, failure::Error> {
+pub fn main(options: CompilerOptions) -> Result<Option<PathBuf>> {
```
Comment (Wodann) [inline crates/mun_compiler/src/lib.rs:61]:
```suggestion
pub fn main(options: CompilerOptions) -> Result<Option<PathBuf>, anyhow::Error> {
```

File: crates/mun_compiler/src/lib.rs [hunk 9]
```diff
@@ -15,6 +15,7 @@ pub use crate::driver::{Config, Driver};
 pub use annotate::{AnnotationBuilder, SliceBuilder, SnippetBuilder};
 pub use mun_codegen::OptimizationLevel;
 
+use anyhow::Result;
```
Comment (Wodann) [inline crates/mun_compiler/src/lib.rs:18]:
```suggestion
```

File: crates/mun_compiler/src/driver.rs [hunk 10]
```diff
@@ -2,6 +2,7 @@
 //! from previous compilation.
 
 use crate::{db::CompilerDatabase, diagnostics::diagnostics, PathOrInline};
+use anyhow::Result;
```
Comment (Wodann) [inline crates/mun_compiler/src/driver.rs:5]:
```suggestion
```

File: crates/mun_compiler_daemon/src/lib.rs [hunk 11]
```diff
@@ -1,13 +1,13 @@
 use std::sync::mpsc::channel;
 use std::time::Duration;
 
-use failure::Error;
+use anyhow::Result;
```
Comment (Wodann) [inline crates/mun_compiler_daemon/src/lib.rs:4]:
```suggestion
```

File: crates/mun_compiler_daemon/src/lib.rs [hunk 12]
```diff
@@ -1,13 +1,13 @@
...
 use mun_compiler::{CompilerOptions, Driver, PathOrInline};
 use notify::{RecommendedWatcher, RecursiveMode, Watcher};
 
 use std::io::stderr;
 
-pub fn main(options: CompilerOptions) -> Result<(), Error> {
+pub fn main(options: CompilerOptions) -> Result<()> {
```
Comment (Wodann) [inline crates/mun_compiler_daemon/src/lib.rs:10]:
```suggestion
pub fn main(options: CompilerOptions) -> Result<(), anyhow::Error> {
```

File: crates/mun_runtime/src/assembly.rs [hunk 13]
```diff
@@ -3,6 +3,7 @@ use std::path::{Path, PathBuf};
 
 use crate::DispatchTable;
 use abi::AssemblyInfo;
+use anyhow::Result;
```
Comment (Wodann) [inline crates/mun_runtime/src/assembly.rs:6]:
```suggestion
```

File: crates/mun_runtime/src/assembly.rs [hunk 14]
```diff
@@ -28,7 +29,7 @@ impl Assembly {
         library_path: &Path,
         gc: Arc<GarbageCollector>,
         runtime_dispatch_table: &DispatchTable,
-    ) -> Result<Self, failure::Error> {
+    ) -> Result<Self> {
```
Comment (Wodann) [inline crates/mun_runtime/src/assembly.rs:32]:
```suggestion
    ) -> Result<Self, anyhow::Error> {
```

File: crates/mun_runtime/src/assembly.rs [hunk 15]
```diff
@@ -150,7 +151,7 @@ impl Assembly {
         &mut self,
         library_path: &Path,
         runtime_dispatch_table: &mut DispatchTable,
-    ) -> Result<(), failure::Error> {
+    ) -> Result<()> {
```
Comment (Wodann) [inline crates/mun_runtime/src/assembly.rs:154]:
```suggestion
    ) -> Result<(), anyhow::Error> {
```

File: crates/mun_runtime/src/assembly/temp_library.rs [hunk 16]
```diff
@@ -1,7 +1,7 @@
 use std::fs;
 use std::path::Path;
 
-use failure::Error;
+use anyhow::Result;
```
Comment (Wodann) [inline crates/mun_runtime/src/assembly/temp_library.rs:4]:
```suggestion
```

File: crates/mun_runtime/src/assembly/temp_library.rs [hunk 17]
```diff
@@ -16,7 +16,7 @@ pub struct TempLibrary {
 }
 
 impl TempLibrary {
-    pub fn new(path: &Path) -> Result<Self, Error> {
+    pub fn new(path: &Path) -> Result<Self> {
```
Comment (Wodann) [inline crates/mun_runtime/src/assembly/temp_library.rs:19]:
```suggestion
    pub fn new(path: &Path) -> Result<Self, anyhow::Error> {
```

File: crates/mun_runtime/src/lib.rs [hunk 18]
```diff
@@ -13,7 +13,7 @@ mod marshal;
 mod reflection;
 mod struct_ref;
 
-use failure::Error;
+pub use anyhow::Result;
```
Comment (Wodann) [inline crates/mun_runtime/src/lib.rs:16]:
```suggestion
```

File: crates/mun_runtime/src/lib.rs [hunk 19]
```diff
@@ -86,7 +86,7 @@ impl RuntimeBuilder {
     }
 
     /// Spawns a [`Runtime`] with the builder's options.
-    pub fn spawn(self) -> Result<Rc<RefCell<Runtime>>, Error> {
+    pub fn spawn(self) -> Result<Rc<RefCell<Runtime>>> {
```
Comment (Wodann) [inline crates/mun_runtime/src/lib.rs:89]:
```suggestion
    pub fn spawn(self) -> Result<Rc<RefCell<Runtime>>, anyhow::Error> {
```

File: crates/mun_runtime/src/lib.rs [hunk 20]
```diff
@@ -202,7 +202,7 @@ impl Runtime {
     /// Constructs a new `Runtime` that loads the library at `library_path` and its
     /// dependencies. The `Runtime` contains a file watcher that is triggered with an interval
     /// of `dur`.
-    pub fn new(mut options: RuntimeOptions) -> Result<Runtime, Error> {
+    pub fn new(mut options: RuntimeOptions) -> Result<Runtime> {
```
Comment (Wodann) [inline crates/mun_runtime/src/lib.rs:205]:
```suggestion
    pub fn new(mut options: RuntimeOptions) -> Result<Runtime, anyhow::Error> {
```

File: crates/mun_runtime/src/lib.rs [hunk 21]
```diff
@@ -234,7 +234,7 @@ impl Runtime {
     }
 
     /// Adds an assembly corresponding to the library at `library_path`.
-    fn add_assembly(&mut self, library_path: &Path) -> Result<(), Error> {
+    fn add_assembly(&mut self, library_path: &Path) -> Result<()> {
```
Comment (Wodann) [inline crates/mun_runtime/src/lib.rs:237]:
```suggestion
    fn add_assembly(&mut self, library_path: &Path) -> Result<(), anyhow::Error> {
```

File: crates/mun_runtime/tests/util.rs [hunk 22]
```diff
@@ -1,5 +1,6 @@
 #![allow(dead_code, unused_macros)]
 
+use anyhow::Result;
```
Comment (Wodann) [inline crates/mun_runtime/tests/util.rs:3]:
```suggestion
```

File: crates/mun_runtime/tests/util.rs [hunk 23]
```diff
@@ -22,7 +23,7 @@ enum RuntimeOrBuilder {
 }
 
 impl RuntimeOrBuilder {
-    pub fn spawn(&mut self) -> Result<(), failure::Error> {
+    pub fn spawn(&mut self) -> Result<()> {
```
Comment (Wodann) [inline crates/mun_runtime/tests/util.rs:26]:
```suggestion
    pub fn spawn(&mut self) -> Result<(), anyhow::Error> {
```

File: crates/mun_runtime/tests/util.rs [hunk 24]
```diff
@@ -71,7 +72,7 @@ impl TestDriver {
     }
 
     /// Spawns a `Runtime` from the `RuntimeBuilder`, if it hadn't already been spawned.
-    pub fn spawn(&mut self) -> Result<(), failure::Error> {
+    pub fn spawn(&mut self) -> Result<()> {
```
Comment (Wodann) [inline crates/mun_runtime/tests/util.rs:75]:
```suggestion
    pub fn spawn(&mut self) -> Result<(), anyhow::Error> {
```

File: crates/mun_target/Cargo.toml [hunk 25]
```diff
@@ -10,7 +10,9 @@ description = "Describes compilation targets for Mun"
 
 [dependencies]
 log = "0.4.8"
-failure = "0.1.7"
+anyhow = "1.0"
+thiserror = "1.0.17"
+displaydoc = "0.1"
```
Comment (Wodann) [inline crates/mun_target/Cargo.toml:15]: Please remove `displaydoc`

File: crates/tools/src/abi.rs [hunk 26]
```diff
@@ -1,5 +1,6 @@
-use crate::{project_root, reformat, update, Result};
-use failure::format_err;
+use crate::{project_root, reformat, update};
+use anyhow::anyhow;
+use anyhow::Result;
```
Comment (Wodann) [inline crates/tools/src/abi.rs:3]:
```suggestion
```

File: crates/tools/src/lib.rs [hunk 27]
```diff
@@ -1,4 +1,4 @@
-pub type Result<T> = std::result::Result<T, failure::Error>;
+use anyhow::Result;
```
Comment (Wodann) [inline crates/tools/src/lib.rs:1]:
```suggestion
```

Comment (Wodann) [reply crates/tools/src/lib.rs:1]: You'll need to replace `Result<T>` with `Result<T, anyhow::Error>` too. This is best for consistency throughout the entire codebase.

File: crates/tools/src/main.rs [hunk 28]
```diff
@@ -1,6 +1,7 @@
 use clap::{App, SubCommand};
 
-use tools::{Overwrite, Result};
+use anyhow::Result;
```
Comment (Wodann) [inline crates/tools/src/main.rs:3]:
```suggestion
```

File: crates/tools/src/runtime_capi.rs [hunk 29]
```diff
@@ -1,4 +1,5 @@
-use crate::{project_root, update, Result};
+use crate::{project_root, update};
+use anyhow::Result;
```
Comment (Wodann) [inline crates/tools/src/runtime_capi.rs:2]:
```suggestion
```

File: crates/tools/src/syntax.rs [hunk 30]
```diff
@@ -1,4 +1,6 @@
-use crate::{project_root, reformat, update, Result};
+use crate::{project_root, reformat, update};
+use anyhow::anyhow;
+use anyhow::Result;
```
Comment (Wodann) [inline crates/tools/src/syntax.rs:3]:
```suggestion
```

File: crates/mun/Cargo.toml (hunk 1 above)
Comment (attliaLin) [reply crates/mun/Cargo.toml:15]: ok！

File: crates/mun_codegen/src/code_gen/linker.rs [hunk 31]
```diff
@@ -1,31 +1,17 @@
...
-use std::fmt;
 use std::path::{Path, PathBuf};
 
-#[derive(Fail, Debug)]
+use thiserror::Error;
+
+#[derive(Error, Debug)]
 pub enum LinkerError {
-    /// Error emitted by the linker
+    #[error("{0}")]
```
Comment (Wodann) [inline crates/mun_codegen/src/code_gen/linker.rs:9]:
```suggestion
    /// Error emitted by the linker
    #[error("{0}")]
```

File: crates/mun_codegen/src/code_gen/linker.rs [hunk 32]
```diff
@@ -1,31 +1,17 @@
...
     LinkError(String),
-
-    /// Error in path conversion
+    #[error("path contains invalid UTF-8 characters: {0:?}")]
```
Comment (Wodann) [inline crates/mun_codegen/src/code_gen/linker.rs:11]:
```suggestion
    /// Error in path conversion
    #[error("path contains invalid UTF-8 characters: {0:?}")]
```

File: crates/mun_runtime/tests/util.rs (hunk 23 above)
Comment (Wodann) [reply crates/mun_runtime/tests/util.rs:25]: this request is still outstanding

File: crates/mun_runtime/tests/util.rs (hunk 24 above)
Comment (Wodann) [reply crates/mun_runtime/tests/util.rs:74]: same

**Comment by Wodann** [comment]
@attliaLin another PR has been created to resolve issue #177. Do you think you can resolve the remaining issues in your PR soon, or would it be better to accept PR #213 instead?

**Comment by attliaLin** [comment]
use #213 is fine.

**Comment by Wodann** [comment]
Thank you for the quick response and your consent.

If you need some help learning how to `git rebase` in the future, I am willing to help. It'd be a shame to lose community members as a result of something like that 🙂 

Closing this PR in favour of #213.

**Pull request closed without merging**